tls/cfsslscan_handshake.go

```diff
@@ -22,9 +71,51 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
 	if err != nil {
 		return
 	}
+ 	if IsECcipherSuite(serverHello.cipherSuite) {
+		// Prime the connection, if necessary, for key
+		// exchange messages by reading off the certificate
+		// message and, if necessary, the OCSP stapling
+		// message
+		var msg interface{}
+		msg, err = c.readHandshake()
+		if err != nil {
+			return
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
+			msg, err = c.readHandshake()
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
+			if !ok {
+				c.sendAlert(alertUnexpectedMessage)
+				return
+			}
+		}
+		
+		var skx *serverKeyExchangeMsg
+ 		skx, err = c.exchangeKeys()
+		if (skx.raw[0] != typeServerKeyExchange) {
+			panic("Bad handshake! Did not get ServerKeyExchange when expected!")
```

Comment:
I was trying to make extra-double sure that I was parsing the right message (while trying to debug the unstable scan results) and this method seemed more readable than the awful reflect-based things you have to do to get a type assertion about the value of a pointer. But I've replaced it with an error and convinced myself that (a) I really am parsing ServerKeyExchange (b) I'm parsing it correctly (or at least consistently with processServerKeyExchange in key_agreement.go) and (c) this check is basically redundant given the type assertion and the logic that checks the message's magic number in readHandshake (see Conn.go).

For now, I've replaced the panic with an error, but it probably makes sense to remove the check altogether, in fact.